### DISCUSSION TITLE
Extremely inefficent tiling code.

### DISCUSSION THREAD

**Discussion by BlackMoons** [FakeFishGames/Barotrauma#12927] [Ideas and suggestions]
Hi. While searching out the cause of low performance drawing in barotrauma, I came across this code here:

https://github.com/Regalis11/Barotrauma/blob/de4726b70ae651f3cfb8e7d7888948f4df180cad/Barotrauma/BarotraumaClient/ClientSource/Sprite/Sprite.cs#L333C13-L333C45

Its a 2 dimensional for loop to render tiled graphics. Meaning 1 draw call per tile. This is really the worst possible way to do it.

Ideally, you'd use a shader to UV repeat the textures, and you can just feed the shader the min/max UV coords for repeat(Do your DDX/DDY math in the shader itself, or pick a mipmap level explicitly when sampling to avoid texture errors on texture wrap edges) if you are doing an atlas

If you are using a whole texture, just use regular UV repeat texture filtering mode on your texture sampler.

This change will drastically reduce draw call count on all submarines since repeat textures are used for hulls, wire conduit holders, walls, floors, platforms, etc.

**Comment by duarm** [comment]
Not sure, but by looking at the code, each sprite draw seems batched, so you would only get 1 draw call, at least up until the size of the batch. Repeating with the sampler would still be better tho.

**Reply by BlackMoons** [reply]
I would hope so, but when making any object that uses many tiles, it seems to cause *massive slowdown* once you get enough tiling going on. (Ie, make a large image with very low scale), so it seems like very little if any batching is actually going on for some reason, or some other overhead exists in batching/sorting that is dependent on number of sprites drawn.
